Using onConnect with SinglePool and context leads to panic

Decided to open new isssue because old one can't be reopened.

Minimal example to reproduce on v9.1.6
```go
package main

import (
	"context"
	"flag"
	"time"

	"github.com/go-pg/pg/v9"
)

func main() {
	pgDSN := flag.String("pg-dsn", "postgres://postgres:pass@localhost:5432/postgres?sslmode=disable", "Postgres connection string")
	flag.Parse()

	pgOpts, err := pg.ParseURL(*pgDSN)
	if err != nil {
		panic(err)
	}

	pgOpts.OnConnect = func(conn *pg.Conn) error {
		_, err := conn.Exec("SELECT pg_sleep(10)")
		return err
	}
	db := pg.Connect(pgOpts)

	// context.WithTimeout/WithDeadline sets i/o deadline and error can't be reproduced
	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		time.Sleep(500 * time.Millisecond)
		cancel()
	}()

	_, err = db.ExecContext(ctx, "SELECT 1")
	if err != nil {
		panic(err)
	}
}
```
Prints 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1143929]

goroutine 18 [running]:
github.com/go-pg/pg/v9/internal/pool.(*SingleConnPool).NewConn(0xc000208090, 0x14512c0, 0xc00001e0b0, 0x0, 0xc00017c688, 0xc00017c738)
       /sources/go/pkg/mod/github.com/go-pg/pg/v9@v9.1.6/internal/pool/pool_single.go:66 +0x29
github.com/go-pg/pg/v9.(*baseDB).cancelRequest(0xc0002000a0, 0x6ac4cb8c00000069, 0x0, 0x0)
       /sources/go/pkg/mod/github.com/go-pg/pg/v9@v9.1.6/base.go:482 +0x76
github.com/go-pg/pg/v9.(*baseDB).withConn.func1(0xc000236000, 0x1451280, 0xc0000212c0, 0xc0002000a0, 0xc000200050)
       /sources/go/pkg/mod/github.com/go-pg/pg/v9@v9.1.6/base.go:145 +0x105
created by github.com/go-pg/pg/v9.(*baseDB).withConn
       /sources/go/pkg/mod/github.com/go-pg/pg/v9@v9.1.6/base.go:141 +0x1e9
```

_Originally posted by @xakep666 in https://github.com/go-pg/pg/issues/1353#issuecomment-671862261_

As you can see this happens when context gets cancelled when query runs in _OnConnect_

Possible workaround: add query hook that always returns `context.Background()` in `BeforeQuery`
```go
type BackgroundContextHook struct {}

func (BackgroundContextHook) BeforeQuery(ctx context.Context, event *pg.QueryEvent) (context.Context, error) {
	return context.Background(), nil
}

func (BackgroundContextHook) AfterQuery(ctx context.Context, event *pg.QueryEvent) error {
	return nil
} 

db.AddQueryHook(BackgroundContextHook{})
```

--- a/base.go
+++ b/base.go
@@ -120,7 +120,7 @@ func (db *baseDB) initConn(ctx context.Context, cn *pool.Conn) error {
 	}
 
 	if db.opt.OnConnect != nil {
-		p := pool.NewSingleConnPool(nil)
+		p := pool.NewSingleConnPool(db.pool)
 		p.SetConn(cn)
 		return db.opt.OnConnect(ctx, newConn(ctx, db.withPool(p)))
 	}

--- a/db_test.go
+++ b/db_test.go
@@ -216,6 +216,28 @@ func TestBigColumn(t *testing.T) {
 	}
 }
 
+var _ = Describe("OnConnect error", func() {
+	It("does not panic", func() {
+		opt := pgOptions()
+		opt.OnConnect = func(ctx context.Context, conn *pg.Conn) error {
+			_, err := conn.Exec("SELECT pg_sleep(10)")
+			return err
+		}
+
+		db := pg.Connect(opt)
+		defer db.Close()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			cancel()
+		}()
+
+		_, err := db.ExecContext(ctx, "SELECT 1")
+		Expect(err).To(MatchError("ERROR #57014 canceling statement due to user request"))
+	})
+})
+
 var _ = Describe("DB", func() {
 	var db *pg.DB
 	var tx *pg.Tx

